// obil-utils/config/index.js
var _ = require('lodash');
var ENV = process.env;


/**
 * @namespace {Object} config Configuration for Utils module
 */
const config = {};


/**
 * Common - default configuration block
 * @memberOf config
 * @property {Object} store Redis configuration @see (@link https://github.com/tj/connect-redis)
 * @property {Object} storeClient Redis client configuration @see (@link https://github.com/mranney/node_redis)
 */
config.common = {
  store: {
    host: 'redis1-19001-stg-chidc2.chidc2.outbrain.com',
    port: 6379,
    database: 1
  },
  storeClient: {
    connect_timeout: 9000, // 9 secs.
    ttl: (60 * 60 * 24 * 30) // Set in secs (30 days).
  }
};


/**
 * Development
 * @memberOf config
 * @property {Object} store Redis configuration @see (@link https://github.com/tj/connect-redis)
 * @property {Object} client Redis client configuration @see (@link https://github.com/mranney/node_redis)
 */
config.development = {
  store: {
    host: '127.0.0.1'
  }
};


/**
 * Jenkins
 * @memberOf config
 * @property {Object} store Redis configuration
 * @property {Object} client Redis client configuration
 */
config.jenkins = {};


/**
 * Staging
 * @memberOf config
 * @property {Object} store Redis configuration
 * @property {Object} client Redis client configuration
 */
config.staging = {};


/**
 * Production
 * @memberOf config
 * @property {Object} store Redis configuration
 * @property {Object} client Redis client configuration
 */
config.production = {};

module.exports = _.merge(config.common, config[(ENV.NODE_ENV || 'development')]);
