// obil-server/lib/routes/helpers/acl.js
/**
 * @class ACL
 *
 * @desc  ACL class constructor.
 */
function Acl() {
  var _store = null;
  var _model = null;

  Object.defineProperties(this, {
    store: {
      get: function() {
        return _store;
      },
      set: function(state) {
        _store = state;
      }
    },
    model: {
      get: function() {
        return _model;
      },
      set: function(state) {
        _model = state;
      }
    }
  });
}


/**
 * @function toString
 *
 * @desc Name of the module
 *
 * @returns <String> name of module
 */
Acl.prototype.toString = function toString() {
  return 'acl';
};


/**
 * @function getUserId
 *
 * @desc Returns user identifier from session (userId).
 * Used to overwrite default ACL userId lookup (@link https://github.com/OptimalBits/node_acl#middleware-numpathcomponents-userid-permissions-)
 *
 * @param {Object} req Request
 * @param {Object} res Response
 *
 * @returns {String} userId
 */
Acl.prototype.getUserId = function getUserId(req) {
  return req.session.userId;
};


module.exports = Acl;
